### PR TITLE
Checked in code to some branch directly

### PR DESCRIPTION
**Pull request by sima-zhu**
Issue: https://github.com/Azure/azure-sdk/issues/2470
Testing pipelines:
Java:https://dev.azure.com/azure-sdk/internal/_build/results?buildId=803251&view=logs&j=df6dbc9d-9cb6-55ac-00c8-bc457d14297a&t=c8a16bce-d539-5a96-ea44-10296799aa9d
.NET:https://dev.azure.com/azure-sdk/internal/_build/results?buildId=804860&view=logs&j=5ccebfdc-da53-51d4-fecb-9480b00e8628&t=d1513c7c-1e34-592b-a96a-fb941417aae9
Python:https://dev.azure.com/azure-sdk/internal/_build/results?buildId=803264&view=logs&j=ceb84d66-6af3-5d95-9875-cd812f26b59e&t=34564ed5-4e54-52a5-372d-3ee518ce2d7c
Js:https://dev.azure.com/azure-sdk/internal/_build/results?buildId=803253&view=logs&j=ceb84d66-6af3-5d95-9875-cd812f26b59e&t=8c18c00a-8976-5c3f-c4f5-8f07ce49366c

### CODE REVIEW

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 1]
```diff
@@ -124,16 +118,16 @@ steps:
...
-      PRBranchName: ${{ parameters.PRBranchName }}-ci
-      CommitMsg: "CI Update for release of ${{ parameters.ArtifactName }}"
-      PRTitle: "Docs.MS CI Updates for ${{ parameters.ArtifactName }}"
-      BaseBranchName: ${{ parameters.OnboardingBranch }}
-      WorkingDirectory: ${{ parameters.WorkingDirectory }}/repo
-      ScriptDirectory: ${{ parameters.WorkingDirectory }}/${{ parameters.ScriptDirectory }}
-      GHReviewersVariable: ${{ parameters.GHReviewersVariable }}
-      GHTeamReviewersVariable: ${{ parameters.GHTeamReviewersVariable }}
-      CloseAfterOpenForTesting: ${{ parameters.CloseAfterOpenForTesting }}
+  # - template: /eng/common/pipelines/templates/steps/create-pull-request.yml
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:121]: I assume you plan to figure out what we want to do with this entire branch of code.

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  PROwner: azure-sdk
+  CommitMsg: not-specified
+  RepoOwner: Azure
+  RepoName: $(Build.Repository.Name)
+  PushArgs:
+  WorkingDirectory: $(System.DefaultWorkingDirectory)'
+  ScriptDirectory: eng/common/scripts
+  SkipCheckingForChanges: false
+
+steps:
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/git-push-changes.yml:12]: Just to close the conversation I agree with splitting this into its own template.

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  PRBranchName: not-specified
+  PROwner: azure-sdk
+  CommitMsg: not-specified
+  RepoOwner: Azure
+  RepoName: $(Build.Repository.Name)
+  PushArgs:
+  WorkingDirectory: $(System.DefaultWorkingDirectory)'
+  ScriptDirectory: eng/common/scripts
+  SkipCheckingForChanges: false
+  RemoteName: not-specified
```
Comment (danieljurek) [inline eng/common/pipelines/templates/steps/git-push-changes.yml:11]: This should default to `azure-sdk-fork` (the default in `git-branch-push.ps1`) so calls to this template from `create-pull-request.yml` (which doesn't have a `RemoteName` parameter) don't change behavior when this checks in.

**Comment by danieljurek** [comment]
Was chatting with @scbedd today about what the world would look like after we implement this change. We agreed that there is the potential for many releases at once to create merge conflicts (especially when simultaneously adding things to the same location in the same file many times concurrently. However, the `git-branch-push.ps1` script handles this scenario with some tenacity. 

The next release will be a good test of that retry loop. In cases where the push step fails despite the retry loop, simply re-running the job will likely result in success. 

In the case of an Onboarding branch we have 2 pushes and a partial success (one push succeeds, other fails) may result in noop pushes in retries but `git-push-changes.yml` is already checking for those changes. In short: I think that scenario is covered as well. 

Our advice to anyone who sees failures in the docs.ms release job should be "let us know that it's happening so we can make adjustments to the implementation logic, retry the job once and it should succeed"

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 4]
```diff
@@ -82,20 +82,15 @@ steps:
...
-    CommitMsg: "Update docs metadata and targeting for release of ${{ parameters.ArtifactName }}"
-    PRTitle: "Docs.MS Release Updates for ${{ parameters.ArtifactName }}"
-    BaseBranchName: ${{ parameters.SourceBranchName }}
     WorkingDirectory: ${{ parameters.WorkingDirectory }}/repo
     ScriptDirectory: ${{ parameters.WorkingDirectory }}/${{ parameters.ScriptDirectory }}
-    GHReviewersVariable: ${{ parameters.GHReviewersVariable }}
-    GHTeamReviewersVariable: ${{ parameters.GHTeamReviewersVariable }} 
-    PRLabels: ${{ parameters.PRLabels }}
-    CloseAfterOpenForTesting: ${{ parameters.CloseAfterOpenForTesting }}
+    RemoteName: 'origin'
```
Comment (danieljurek) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:93]: We should either keep this as `origin` and update the `git remote rm azure-sdk-fork` in the `OnboardingBranch` scenario on line 100 below, or set this as `azure-sdk-fork` here and on line 129 in this file as well.

**Comment by weshaggard** [comment]
> We agreed that there is the potential for many releases at once to create merge conflicts (especially when simultaneously adding things to the same location in the same file many times concurrently. However, the git-branch-push.ps1 script handles this scenario with some tenacity.

Do we really think there will be that many conflicts? There might be other changes but hopefully not in the same file. If there are a lot of changes to the same files we might need to see if we can find a way to potentially separate them in some way to avoid conflicts. The retry logic will help for concurrent pushes and some basic conflicts but if there are files touching lines next to each other in the same files this retry logic will start to fall apart.

**Comment by sima-zhu** [comment]
> Do we really think there will be that many conflicts? There might be other changes but hopefully not in the same file. If there are a lot of changes to the same files we might need to see if we can find a way to potentially separate them in some way to avoid conflicts. The retry logic will help for concurrent pushes and some basic conflicts but if there are files touching lines next to each other in the same files this retry logic will start to fall apart.

Every release build has to touch the `package.json` file. We got conflicts even in PRs. I am also looking for the github API which can lock the branch when we checked in the changes. The next incoming push operation has to pull the latest before push

**Comment by weshaggard** [comment]
>I am also looking for the github API which can lock the branch when we checked in the changes. 

There are no such locking that I'm aware of and even if there was I would avoid trying to use it. As long as the likelihood of 2 jobs publishing to that file at the exact same time is pretty low we might be OK. HOwever if we can figure out some way to format the package.json file so that no 2 edits would be on exactly the same adjacent lines (i.e. maybe we could have some spacing or formatting between the often edit lines) then the retry would still mostly work.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 3 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: If it calls create-pull-request, then `RemoteName` here did not get set. The value will default to `azure-sdk-fork` to `git-branch-push.ps1`, right? This is the same behavior as before.

File: eng/common/scripts/git-branch-push.ps1 [hunk 5]
```diff
@@ -60,8 +60,14 @@ if ($LASTEXITCODE -ne 0)
     exit $LASTEXITCODE
 }
 
-Write-Host "git checkout -b $PRBranchName"
-git checkout -b $PRBranchName
+# Check if the branch is default branch in remote.
+$defaultBranch = (git remote show $RemoteName | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:64]: Why is this check needed here? I assume we can checkout the branch still even if it is the default branch.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 6]
```diff
@@ -82,27 +84,21 @@ steps:
...
-    CommitMsg: "Update docs metadata and targeting for release of ${{ parameters.ArtifactName }}"
-    PRTitle: "Docs.MS Release Updates for ${{ parameters.ArtifactName }}"
-    BaseBranchName: ${{ parameters.SourceBranchName }}
     WorkingDirectory: ${{ parameters.WorkingDirectory }}/repo
     ScriptDirectory: ${{ parameters.WorkingDirectory }}/${{ parameters.ScriptDirectory }}
-    GHReviewersVariable: ${{ parameters.GHReviewersVariable }}
-    GHTeamReviewersVariable: ${{ parameters.GHTeamReviewersVariable }} 
-    PRLabels: ${{ parameters.PRLabels }}
-    CloseAfterOpenForTesting: ${{ parameters.CloseAfterOpenForTesting }}
+    #RemoteName: ${{ parameters.RemoteName }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:95]: Why is this commended out?

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    pwsh: true
+    workingDirectory: ${{ parameters.WorkingDirectory }}
+    filePath: ${{ parameters.ScriptDirectory }}/git-branch-push.ps1
+    arguments: >
+      -PRBranchName "${{ parameters.PRBranchName }}"
+      -CommitMsg "${{ parameters.CommitMsg }}"
+      -GitUrl "https://$(azuresdk-github-pat)@github.com/${{ parameters.PROwner }}/$(RepoNameWithoutOwner).git"
+      -PushArgs "${{ parameters.PushArgs }}"
+      -SkipCommit $${{ parameters.SkipCheckingForChanges }}
+      -RemoteName ${{ parameters.RemoteName }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/git-push-changes.yml:56]: If you expect this value to sometimes be empty you need to surround it with quotes so the empty case looks like `-RemoteName ""`

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 3 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: Currently you would be passing not-specified as the value which is not correct. You should either leave it empty or default it to a valid value as Daniel suggests.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml (hunk 6 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:95]: That's only for testing purpose. I will add it back.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 3 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: Sure. I am having a test, and make sure the default value is correct if it is not setting.

File: eng/common/scripts/git-branch-push.ps1 (hunk 5 above)
Comment (sima-zhu) [reply eng/common/scripts/git-branch-push.ps1:64]: `git checkout -b master` will fail if $PRBranchName is default branch.

Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:64]: I don't think that is just for the default branch I suspect that is true for any case where you are already on the PRBranchName so maybe the check should be more geared towards that instead of only the default branch. Or we can also just handle any errors that come when trying to checkout the branch in question.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 8]
```diff
@@ -20,8 +20,10 @@ parameters:
   OnboardingBranch: ''
   CloseAfterOpenForTesting: false
   SkipPackageJson: false
+  RemoteName: 'origin'
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:23]: Lets remove passing a RemoteName in and let it always be created by the script.

File: eng/common/scripts/git-branch-push.ps1 [hunk 9]
```diff
@@ -60,8 +60,14 @@ if ($LASTEXITCODE -ne 0)
...
+Write-Host "The default branch is $defaultBranch."
+if ($defaultBranch -ne $PRBranchName) {
+    Write-Host "git checkout -b $PRBranchName"
+    git checkout -b $PRBranchName
+}
+
 if ($LASTEXITCODE -ne 0)
 {
     Write-Error "Unable to create branch LASTEXITCODE=$($LASTEXITCODE), see command output above."
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:73]: Lets make sure that this script fails if we run out of retries.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 10]
```diff
@@ -33,7 +33,7 @@ steps:
   displayName: Enable Long Paths if Necessary
   
 - pwsh: |
-    git clone https://github.com/${{ parameters.TargetDocRepoOwner }}/${{ parameters.TargetDocRepoName }} ${{ parameters.WorkingDirectory }}/repo
+    git clone https://$(azuresdk-github-pat)@github.com/${{ parameters.TargetDocRepoOwner }}/${{ parameters.TargetDocRepoName }} ${{ parameters.WorkingDirectory }}/repo
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:36]: Lets not do this here. It isn't needed for the clone only the remote which we are going to push.

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  PRBranchName: not-specified
+  PROwner: azure-sdk
+  CommitMsg: not-specified
+  RepoOwner: Azure
+  RepoName: $(Build.Repository.Name)
+  PushArgs:
+  WorkingDirectory: $(System.DefaultWorkingDirectory)'
+  ScriptDirectory: eng/common/scripts
+  SkipCheckingForChanges: false
+  RemoteName: ''
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/git-push-changes.yml:11]: Lets completely remove this as a parameter as I don't think we need it and it will most likely cause more issues then it solves.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 7 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:56]: Lets remove this parameter.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 11 above)
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: I am afraid `create-pull-request` has no way to create the fork if completely remove it

File: eng/common/scripts/git-branch-push.ps1 (hunk 5 above)
Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:64]: Lets remove this change completely unless there is a know issue and if there is I'd be interested in a link to the failures it causes.

File: eng/common/scripts/git-branch-push.ps1 [hunk 12]
```diff
@@ -168,7 +174,7 @@ do
     }
 } while($needsRetry -and $tryNumber -le $numberOfRetries)
 
-if ($LASTEXITCODE -ne 0)
+if ($LASTEXITCODE -ne 0 -or $tryNumber -ge $numberOfRetries)
 {
     Write-Error "Unable to push commit after $($tryNumber) retries LASTEXITCODE=$($LASTEXITCODE), see command output above."
     exit $LASTEXITCODE
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:180]: In the case of the retry fails we need to also return a non-zero exit code.

File: eng/common/pipelines/templates/steps/git-push-changes.yml (hunk 11 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: It doesn't actually create the fork it only creates the remote name. The fork or repo you are pushing to is controlled by the GitUrl that is created,

Comment (weshaggard) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:11]: Those are controlled by the RepoName and PROwner.

File: eng/common/scripts/git-branch-push.ps1 [hunk 13]
```diff
@@ -50,18 +51,32 @@ if (!(git remote | ? {$_ -eq $RemoteName}))
         Write-Error "Unable to add remote LASTEXITCODE=$($LASTEXITCODE), see command output above."
         exit $LASTEXITCODE
     }
+    Write-Host "git fetch $RemoteName"
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:54]: I don't understand why this is needed? Also I don't think this script should be in the business of doing clone directly like you are doing below. Can you help me understand the scenario you are after?

File: eng/common/scripts/git-branch-push.ps1 [hunk 14]
```diff
@@ -168,8 +187,12 @@ do
     }
 } while($needsRetry -and $tryNumber -le $numberOfRetries)
 
-if ($LASTEXITCODE -ne 0)
+if ($LASTEXITCODE -ne 0 -or $tryNumber -ge $numberOfRetries)
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:190]:
```suggestion
if ($LASTEXITCODE -ne 0 -or $tryNumber -gt $numberOfRetries)
```

File: eng/pipelines/templates/steps/sync-directory.yml [hunk 15]
```diff
@@ -53,6 +53,7 @@ steps:
           -GitUrl "https://$(azuresdk-github-pat)@github.com/${{ parameters.PROwner }}/${{ repo }}.git"
           -PushArgs "${{ parameters.PushArgs }}"
           -SkipCommit $${{parameters.SkipCheckingForChanges}}
+          -RemoteName 'azure-sdk-fork'
```
Comment (weshaggard) [inline eng/pipelines/templates/steps/sync-directory.yml:56]: We shouldn't need to pass this through as the script already defaults to this value.

File: eng/common/scripts/git-branch-push.ps1 [hunk 16]
```diff
@@ -98,13 +113,18 @@ $tryNumber = 0
 do
 {
     $needsRetry = $false
+    if (!$RemoteName)
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:116]: Why is this needed?

File: eng/common/scripts/git-branch-push.ps1 [hunk 17]
```diff
@@ -27,7 +27,8 @@ param(
     [Parameter(Mandatory = $false)]
     [string] $PushArgs = "",
 
-    [string] $RemoteName = "azure-sdk-fork",
+    [Parameter(Mandatory = $false)]
+    [string] $RemoteName = "",
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:31]: You should be keeping the default remote name I don't understand a scenario where we ever need pass a different name. We might even be able to completely remove this parameter because I don't believe we should ever need it except for is some rare care were we want to use an exact remote we already set. I don't believe we should ever really use origin here. As we saw unless the remote was set up the exact matching giturl this script will fail so we might as well always create a new remote that exactly matches the giturl.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 18]
```diff
@@ -34,12 +33,12 @@ steps:
...
 - pwsh: |
     git clone https://github.com/${{ parameters.TargetDocRepoOwner }}/${{ parameters.TargetDocRepoName }} ${{ parameters.WorkingDirectory }}/repo
-
+ 
     try {
       Push-Location ${{ parameters.WorkingDirectory }}/repo
-
-      Write-Host "git checkout ${{ parameters.SourceBranchName }}"
-      git checkout ${{ parameters.SourceBranchName }}
+      $setDefaultBranch = (git remote show origin | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:39]: Why not use the common template to get the default branch?

Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/docs-metadata-release.yml:39]: Yes, will do

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 19]
```diff
@@ -82,27 +81,20 @@ steps:
...
+    CommitMsg: "Update docs metadata and targeting for release of ${{ parameters.ArtifactName }}"
     RepoName: ${{ parameters.TargetDocRepoName }}
     RepoOwner: ${{ parameters.TargetDocRepoOwner }}
-    PRBranchName: ${{ parameters.PRBranchName }}
-    CommitMsg: "Update docs metadata and targeting for release of ${{ parameters.ArtifactName }}"
-    PRTitle: "Docs.MS Release Updates for ${{ parameters.ArtifactName }}"
-    BaseBranchName: ${{ parameters.SourceBranchName }}
     WorkingDirectory: ${{ parameters.WorkingDirectory }}/repo
     ScriptDirectory: ${{ parameters.WorkingDirectory }}/${{ parameters.ScriptDirectory }}
-    GHReviewersVariable: ${{ parameters.GHReviewersVariable }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:95]: These template parameters should be removed (assuming we can do that without breaking things). Also we should remove the steps that are actually gathering those values now that we aren't using them.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 20]
```diff
@@ -82,27 +81,20 @@ steps:
...
-    GHTeamReviewersVariable: ${{ parameters.GHTeamReviewersVariable }} 
-    PRLabels: ${{ parameters.PRLabels }}
-    CloseAfterOpenForTesting: ${{ parameters.CloseAfterOpenForTesting }}
 
 - ${{if ne( parameters['OnboardingBranch'], '')}}:
   - pwsh: |
       Push-Location ${{ parameters.WorkingDirectory }}/repo
 
       git reset --hard HEAD
-      git remote rm azure-sdk-fork
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:105]: This might still be needed if we run this more then once in the same pipeline.

File: eng/common/scripts/git-branch-push.ps1 [hunk 21]
```diff
@@ -27,7 +27,8 @@ param(
     [Parameter(Mandatory = $false)]
     [string] $PushArgs = "",
 
-    [string] $RemoteName = "azure-sdk-fork",
+    # [Parameter(Mandatory = $false)]
+    # [string] $RemoteName = "azure-sdk-fork",
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:31]: This might be a breaking change. Can you verify that we aren't passing this anywhere?

Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:31]: I didn't see any usages of it when it was added https://github.com/Azure/azure-sdk-tools/commit/60121995ebeffb4ebb6115a745eb3cb357cae9c4 and from a quick search https://github.com/Azure/azure-sdk-tools/search?q=RemoteName. So it is probably OK make this change.

File: eng/common/scripts/git-branch-push.ps1 [hunk 22]
```diff
@@ -123,9 +121,8 @@ do
                 Write-Error "Unable to create diff file LASTEXITCODE=$($LASTEXITCODE), see command output above."
                 continue
             }
-
-            Write-Host "git reset --hard $RemoteName/${PRBranchName}"
-            git reset --hard $RemoteName/${PRBranchName}
+            Write-Host "git reset --hard origin/${PRBranchName}"
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:124]: Why did you change this to origin? It should still be the RemoteName as it was before.

File: eng/common/scripts/git-branch-push.ps1 [hunk 23]
```diff
@@ -40,28 +41,24 @@ param(
 # Without explicitly setting the ErrorActionPreference to continue the script
 # would fail the first time git wrote command output.
 $ErrorActionPreference = "Continue"
+$RemoteName = "azure-sdk-fork"
 
-if (!(git remote | ? {$_ -eq $RemoteName}))
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:44]: If the remote exists and doesn't match the GitUrl this script well end up failing. So we should add a check to ensure that the remote name doesn't exists but if it does exist that it matches the GitUrl.

Comment (weshaggard) [reply eng/common/scripts/git-branch-push.ps1:44]: Something like:

```
if ((git remote) -contains $RemoteName)
{
  $remoteUrl = git remote get-url $RemoteName
  if ($remoteUrl -ne $GutUrl)
  {
     Write-Error "Remote with name $RemoteName already exists with an incompatible url [$remoteUrl] which should be [$giturl]."
     exit 1
  }
}

```

File: eng/common/pipelines/templates/steps/create-pull-request.yml [hunk 24]
```diff
@@ -23,49 +23,18 @@ parameters:
...
+    PRBranchName: ${{ parameters.PRBranchName }}
+    PROwner: ${{ parameters.PROwner }}
+    CommitMsg: ${{ parameters.CommitMsg }}
+    RepoOwner: ${{ parameters.RepoOwner }}
+    RepoName: ${{ parameters.RepoName }}
+    PushArgs: ${{ parameters.PushArgs }}
+    WorkingDirectory: ${{ parameters.WorkingDirectory }}
+    ScriptDirectory: ${{ parameters.ScriptDirectory }}
+    SkipCheckingForChanges: ${{ parameters.SkipCheckingForChanges }} 
+    RemoteName: ${{ parameters.RemoteName }}
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/create-pull-request.yml:37]: RemoteName should be removed here.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 25]
```diff
@@ -9,7 +9,6 @@ parameters:
   TargetDocRepoName: ''
   TargetDocRepoOwner: ''
   PRBranchName: 'master-rdme'
-  SourceBranchName: 'master'
   PRLabels: 'auto-merge'
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:12]: After all the work is in go through these all these parameters remove any of the ones we are no longer using.

File: eng/common/pipelines/templates/steps/docs-metadata-release.yml [hunk 26]
```diff
@@ -34,18 +33,17 @@ steps:
   
 - pwsh: |
     git clone https://github.com/${{ parameters.TargetDocRepoOwner }}/${{ parameters.TargetDocRepoName }} ${{ parameters.WorkingDirectory }}/repo
-
+ 
     try {
       Push-Location ${{ parameters.WorkingDirectory }}/repo
-
-      Write-Host "git checkout ${{ parameters.SourceBranchName }}"
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/docs-metadata-release.yml:41]: Why do we not need this checkout anymore? If it isn't needed then we can also remove the try/finally block and the directory changing.

**Review by danieljurek — APPROVED** [review]
One minor naming issue. This process works for our somewhat more nuanced C++ vcpkg release process as evidenced here: https://github.com/microsoft/vcpkg/pull/16893

File: eng/common/scripts/git-branch-push.ps1 [hunk 27]
```diff
@@ -41,27 +42,47 @@ param(
...
+  }
+}
+# Check if the PRBranch is current branch.
+$currentBranch = git branch --show-current
+Write-Host "The current branch is $currentBranch."
+
+if ($currentBranch -ne $PRBranchName) 
+{
+  # Check if the $PRBranch is default branch, if so, we swtich to default one
+  $defaultBranch = (git remote show $RemoteName | Out-String) -replace "(?ms).*HEAD branch: (\w+).*", '$1'
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:72]: Why do we need this logic? We should just always checkout -b $PRBranchName.

File: eng/common/pipelines/templates/steps/git-push-changes.yml [hunk 28]
```diff
@@ -0,0 +1,54 @@
+parameters:
+  PRBranchName: not-specified
+  PROwner: azure-sdk
```
[the remark this answers is not in the document]
Comment (sima-zhu) [reply eng/common/pipelines/templates/steps/git-push-changes.yml:3]: The naming here is unclear.
To me, we have base and target.
`To: targetRepoName/targetRepoBranch <- From: baseRepoName/baseRepoBranch`

In this scenario, `PROwner = baseRepoName`, `PRBranchName = baseRepoBranch`

File: eng/common/scripts/git-branch-push.ps1 [hunk 29]
```diff
@@ -41,27 +42,38 @@ param(
...
+  {
+    Write-Error "Unable to add remote LASTEXITCODE=$($LASTEXITCODE), see command output above."
     exit $LASTEXITCODE
+  }
+}
+# Checkout to PRBranch, create new one if not exists.
+try 
+{
+    Write-Host "git checkout -b $PRBranchName"
+    git checkout -b $PRBranchName
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:69]: I don't believe this will throw an exception you probably need to capture the LASTEXITCODE like other cases.

Comment (sima-zhu) [reply eng/common/scripts/git-branch-push.ps1]: Right. It is not throwing exception.

`(git checkout $PRBranchName | git checkout -b $PRBranchName) 2>&1`
Can we do this and set the `$LASTEXITCODE = 0`?

File: eng/common/scripts/git-branch-push.ps1 [hunk 30]
```diff
@@ -41,32 +42,30 @@ param(
...
+  git remote add $RemoteName $GitUrl
+  if ($LASTEXITCODE -ne 0)
+  {
+    Write-Error "Unable to add remote LASTEXITCODE=$($LASTEXITCODE), see command output above."
     exit $LASTEXITCODE
+  }
 }
+# Checkout to PRBranch, create new one if not exists.
+Write-Host "git checkout $PRBranchName."
+(git checkout $PRBranchName | git checkout -b $PRBranchName) 2>&1
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:67]: This `|` is not going to do what you want it will still always execute both commands. I was testing with something like:

```
# Check to see if the branch exists to a
git show-ref --verify --quiet refs/heads/$PRBranchName
if ($LASTEXITCODE -eq 0) {
  git checkout $PRBranchName 
} 
else {
  git checkout -b $PRBranchName
}
```

File: eng/common/scripts/git-branch-push.ps1 [hunk 31]
```diff
@@ -41,31 +42,34 @@ param(
...
-    exit $LASTEXITCODE
+# Checkout to $PRBranch, create new one if not exists.
+Write-Host "git checkout $PRBranchName."
+git show-ref --verify --quiet refs/heads/$PRBranchName
+if ($LASTEXITCODE -eq 0) {
+  git checkout $PRBranchName 
+} 
+else {
+  git checkout -b $PRBranchName
 }
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:74]: Just for safety we should keep after this block
```
if ($LASTEXITCODE -ne 0)
{
    Write-Error "Unable to create branch LASTEXITCODE=$($LASTEXITCODE), see command output above."
    exit $LASTEXITCODE
}
```

File: eng/common/scripts/git-branch-push.ps1 [hunk 32]
```diff
@@ -41,33 +42,40 @@ param(
...
-    Write-Error "Unable to fetch remote LASTEXITCODE=$($LASTEXITCODE), see command output above."
+  Write-Host "git remote add $RemoteName $GitUrl"
+  git remote add $RemoteName $GitUrl
+  if ($LASTEXITCODE -ne 0)
+  {
+    Write-Error "Unable to add remote LASTEXITCODE=$($LASTEXITCODE), see command output above."
     exit $LASTEXITCODE
+  }
+}
+# Checkout to $PRBranch, create new one if not exists.
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:65]: What happen to your code to check to see if we were already on that branch? I think that is still needed here to ensure those checkout commands don't fail.

File: eng/common/pipelines/templates/steps/set-default-branch.yml [hunk 33]
```diff
@@ -8,4 +8,3 @@ steps:
     echo "##vso[task.setvariable variable=DefaultBranch]$setDefaultBranch"
   displayName: "Setup Default Branch"
   workingDirectory: ${{ parameters.workingDirectory }}
-  condition: eq(variables['DefaultBranch'], '')
```
Comment (sima-zhu) [inline eng/common/pipelines/templates/steps/set-default-branch.yml:11]: Remove this as eng-sync clone every repo and reset defaultBranch whenever in a new repo.  The condition here will keep the defaultBranch as the same one.

File: eng/common/scripts/git-branch-push.ps1 (hunk 32 above)
Comment (sima-zhu) [reply eng/common/scripts/git-branch-push.ps1:65]: Just checked the scenario. 
If we already sit on branch `aBranch`,
run `git show-ref --verify --quiet refs/heads/aBranch` will return 0 for `$LASTEXITCODE`.

Then execute `git checkout $PRBranchName` will return 0 for `$LASTEXITCODE`.

We are good in this case.

File: eng/common/scripts/git-branch-push.ps1 [hunk 34]
```diff
@@ -41,33 +42,40 @@ param(
...
+Write-Host "git checkout $PRBranchName."
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:66]: Just to ensure better diagnostics can you move this logging next to the command that actually runs and then make it match the command.

File: eng/common/pipelines/templates/steps/set-default-branch.yml (hunk 33 above)
Comment (weshaggard) [reply eng/common/pipelines/templates/steps/set-default-branch.yml:11]: True we might need to run this step multiple times in the same pipeline.

**Review by weshaggard — APPROVED** [review]

File: eng/common/scripts/git-branch-push.ps1 [hunk 35]
```diff
@@ -41,33 +42,40 @@ param(
...
+  }
+}
+# Checkout to $PRBranch, create new one if not exists.
+git show-ref --verify --quiet refs/heads/$PRBranchName
+if ($LASTEXITCODE -eq 0) {
+  Write-Host "git checkout $PRBranchName."
+  git checkout $PRBranchName 
+} 
+else {
+  git checkout -b $PRBranchName
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:72]:
```suggestion
  Write-Host "git checkout -b $PRBranchName."
  git checkout -b $PRBranchName
```

File: eng/common/scripts/git-branch-push.ps1 [hunk 36]
```diff
@@ -41,33 +42,40 @@ param(
...
 }
-
-Write-Host "git checkout -b $PRBranchName"
-git checkout -b $PRBranchName
 if ($LASTEXITCODE -ne 0)
 {
     Write-Error "Unable to create branch LASTEXITCODE=$($LASTEXITCODE), see command output above."
     exit $LASTEXITCODE
 }
-
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:70]: nit: Lets try to keep this empty line for spacing.

File: eng/common/scripts/git-branch-push.ps1 [hunk 37]
```diff
@@ -123,7 +132,6 @@ do
                 Write-Error "Unable to create diff file LASTEXITCODE=$($LASTEXITCODE), see command output above."
                 continue
             }
-
```
Comment (weshaggard) [inline eng/common/scripts/git-branch-push.ps1:126]: nit: lets try to keep this empty line.
